### PR TITLE
Add (some) support for custom scopes and reduce number of skill completed notifications

### PR DESCRIPTION
**Pull request by wbSD**
Added a setting for scopes - easy to fill out with "Copy scopes to clipboard" from the EVE Developer Application Settings.
Added a button for auto-filling the old definition.

The way the query monitors were chained after CharacterSheet meant that if just one of them failed, the rest wouldn't load.
Now they are all queried when their parent (CharacterSheet) is queried. Same principle applied to MailingLists.

### CODE REVIEW

**Comment by peterhaneve** [comment]
I was planning on implementing limited scopes in a different way, more similar to how EVEMon originally did it. The copy scopes is interesting, but many lay users of EVEMon will have no idea which scopes to select to get a certain functionality. The reduced scopes should also provide a workaround to users with low limits on max URL length which could be causing a wide variety of reported issues.

The chained queries using the attribute is a really good idea. The problem is when to call `OnCharacterUpdated` / `OnCharacterInfoUpdated` - it needs to execute when **all** of the character sheet information is collected. In your proposed patch, `OnWalletBalanceUpdated` which is the only method calling `OnCharacterUpdated` could execute before some of the other information is fetched, resulting in incomplete or wrong character information.

I initially ran `OnCharacterInfoUpdated` once in every Import method but that workaround exacerbated the "multiple skill trained notifications" problem even more. My current tactic would be to allow the other character sheet pages to be queried even if the first ones failed, but any other suggestions would be helpful.

**Comment by wbSD** [comment]
Valid points.

I've modified the scopes so by default it will fill out the constants (meaning the average user won't have to do anything). Ideally there should be some better way of setting it up, with information about which functionality links to what scope. The main reason for adding that setting was so users that are paranoid about exposing everything will still have a way to use EVEMon with a limited access key.

For the events, I've put it in the timer, subscribing on a successful CharacterSheet update, and unsubscribing once all CharacterSheet child querymonitors are done updating.

As for those multiple skill trained notifications, it looks like the problem is somewhere around here:
https://github.com/peterhaneve/evemon/blob/a4d32873e00776d3bcd54dae7cf42c62721ab2c0/src/EVEMon/MainWindow.cs#L741
The old SkillCompletionNotificationEventArgs aren't removed from m_popupNotifications until 1 minute later, so if another notification is sent before that, the old SkillCompletionNotificationEventArgs is raised again.
If the same SkillCompletionNotificationEventArgs is sent again within that one minute, it becomes duplicated (or more).

File: src/EVEMon.Common/QueryMonitor/CharacterDataQuerying.cs [hunk 1]
```diff
@@ -231,12 +234,45 @@ internal void Dispose()
...
+        }
+
         /// <summary>
         /// Processes the queried character's character sheet information.
         /// </summary>
         /// <param name="result"></param>
         private void OnCharacterSheetUpdated(EsiAPICharacterSheet result)
         {
+            // Subscribe timer tick to wait for the rest of the character sheet operations to finish
+            EveMonClient.TimerTick -= EveMonClient_TimerTick_CharacterSheetUpdated;
```
Comment (peterhaneve) [inline src/EVEMon.Common/QueryMonitor/CharacterDataQuerying.cs]: If you are concerned about double subscription, you should use a Boolean member flag.

**Comment by wbSD** [comment]
I managed to hide this in the last commit message:
Skills list should now show the correct level/sp after both Skills and SkillQueue completed, no matter which order.
Excess skill completed notifications should also be prevented now (as skills aren't imported at lower than actual level).
One thing remains - skill queue import only will import at wrong levels. (meaning it will just bump the skill up x levels from 0)
I'm gonna look at that last thing later today.

**Comment by wbSD** [comment]
Skills are now updated as properly as possible on skill queue import.
Any skills that come from ESI as completed will be bumped to the correct level and SP.
The first uncompleted skill will be bumped to the current level (training - 1) and current SP.

I've been testing some different scenarios (paused queue, 1 skill queue, multiple skill queue with some completed), and it's now only giving me skill completed notifications when a skill actually completes, or when adding a new character that has completed skills still in the queue.

**Comment by wbSD** [comment]
Query monitor/skills: https://github.com/peterhaneve/evemon/pull/12
Scopes: https://github.com/peterhaneve/evemon/pull/13
Closing this one

**Pull request closed without merging**